File: data-structures/java/data-structures/src/main/java/com/data_structures/list/linked_list/doubly_linked_list/DoublyLinkedList.java
package com.data_structures.list.linked_list.doubly_linked_list;


import java.util.HashSet;
import java.util.Iterator;

class DoublyLinkedList<Item> implements Iterable<Item>, IDoublyLinkedList<Item> {

    private DoublyLinkedListNode<Item> head;
    private int size;


    DoublyLinkedList() {
        this.head = null;
        this.size = 0;
    }

    public void insert(Item item) {
        insertNth(item, size);
    }

    public void createLoop() {
        DoublyLinkedListNode<Item> currentNode = this.head;
        while (currentNode.getNext() != null) {
            currentNode = currentNode.getNext();
        }
        currentNode.setNext(this.head);
    }

    public void insertNth(Item item, int position) {
        checkBounds(position, 0, size);
        DoublyLinkedListNode<Item> newNode = new DoublyLinkedListNode<Item>(item);
        if (this.head == null) {
            this.head = newNode;
            this.size++;
            return;
        }
        if (position == 0) {
            newNode.setNext(this.head);
            this.head.setPrev(newNode);
            this.head = newNode;
            this.size++;
            return;
        }

        DoublyLinkedListNode<Item> tempNode = this.head;
        for (int i = 0; i < position - 1; ++i) {
            tempNode = tempNode.getNext();
        }
        newNode.setNext(tempNode.getNext());
        newNode.setPrev(tempNode);
        tempNode.setNext(newNode);
        tempNode.getNext().setPrev(newNode);
        this.size++;
    }

    public void checkBounds(int position, int low, int high) {
        if (position > high || position < low) {
            throw new IndexOutOfBoundsException(position + "");
        }
    }

    public void printItems() {
        Iterator<Item> iterator = iterator();
        while (iterator.hasNext()) {
            Item data = iterator.next();
            System.out.print(data + " => ");
        }
        System.out.println("NULL");
    }

    public DoublyLinkedListIterator<Item> iterator() {
        return new DoublyLinkedListIterator<Item>(this.head);
    }

    //Floyd’s Cycle Finding Algorithm (slow(+1) and fast(+2) pointer)
    public Item middle() {
        DoublyLinkedListNode<Item> temp = this.head;
        DoublyLinkedListNode<Item> slowPtr = temp;
        DoublyLinkedListNode<Item> fastPtr = temp;
        while (slowPtr.getNext() != null && fastPtr.getNext() != null && fastPtr.getNext().getNext() != null) {
            slowPtr = slowPtr.getNext();
            fastPtr = fastPtr.getNext().getNext();
        }
        return slowPtr.getData();
    }

    @Override
    public boolean detectLoopByFloydCycleFindingAlorithm() {
        DoublyLinkedListNode<Item> slowPtr = this.head;
        DoublyLinkedListNode<Item> fastPtr = this.head;
        boolean isLoopDetected = false;
        while (slowPtr.getNext() != null && fastPtr.getNext() != null && fastPtr.getNext().getNext() != null) {
            slowPtr = slowPtr.getNext();
            fastPtr = fastPtr.getNext().getNext();
            if (slowPtr == fastPtr) {
                isLoopDetected = true;
                break;
            }
        }
        return isLoopDetected;
    }

    @Override
    public boolean detectLoopByHash() {
        HashSet<DoublyLinkedListNode<Item>> set = new HashSet<DoublyLinkedListNode<Item>>();
        boolean isLoopDetected = false;
        DoublyLinkedListIterator iterator = iterator();

        while (iterator.hasNext()) {
            DoublyLinkedListNode<Item> currentNode = iterator.nextNode();
            if (set.contains(currentNode)) {
                isLoopDetected = true;
                break;
            } else {
                set.add(currentNode);
            }
        }
        return isLoopDetected;
    }

    public static void main(String[] args) {
        DoublyLinkedList<String> doublyLinkedList = new DoublyLinkedList<String>();
        doublyLinkedList.insert("avinash");
        doublyLinkedList.insert("nishad");
        doublyLinkedList.insert("Demo");
        doublyLinkedList.printItems();
        System.out.println(doublyLinkedList.middle());
        // singlyLinkedList.createLoop();
        System.out.println(doublyLinkedList.detectLoopByFloydCycleFindingAlorithm());
        System.out.println(doublyLinkedList.detectLoopByHash());
    }
}

File: data-structures/python/data-structures/src/list/linked_list/doubly_linked_list/DoublyLinkedList.py
from src.list.linked_list.doubly_linked_list.IDoublyLinkedList import IDoublyLinkedList
from src.list.linked_list.doubly_linked_list.DoublyLinkedListNode import DoublyLinkedListNode


class DoublyLinkedList(IDoublyLinkedList):
    def __init__(self, head=None, tail=None):
        self._head = head
        self._tail = tail

    @property
    def head(self):
        return self._head

    @head.setter
    def head(self, val):
        self._head = val

    @head.deleter
    def head(self):
        pass

    @property
    def tail(self):
        return self._tail

    @tail.setter
    def tail(self, val):
        self._tail = val

    @tail.deleter
    def tail(self):
        del self._tail

    def add_data(self, data):
        newNode = DoublyLinkedListNode(data, None, None)
        if self.head is None:
            self.head = self.tail = newNode
        else:
            tempNode = self.tail
            self.tail.next = newNode
            newNode.prev = tempNode
            self.tail = newNode

    def remove_data(self, data):
        tempHead = self.head
        self.head = None
        self.tail = None
        while tempHead is not None:
            if tempHead.data != data:
                self.add_data(tempHead.data)
            tempHead = tempHead.next

    def print_nodes(self):
        tempRef = self.head
        while tempRef is not None:
            print(tempRef.data)
            tempRef = tempRef.next

    def __iter__(self):
        self.tempNode = self.head
        return self

    def __next__(self):
        currentValue = None
        if self.tempNode and self.tempNode.data:
            currentValue = self.tempNode.data
            self.tempNode = self.tempNode.next
        return currentValue


doublyLinkedList = DoublyLinkedList()
doublyLinkedList.add_data(1)
doublyLinkedList.add_data(2)
doublyLinkedList.add_data(1)
iterator = iter(doublyLinkedList)
print(next(iterator))
print(next(iterator))
print(next(iterator))
